src/app/clusters/power-source-server/power-source-server.h

```diff
@@ -0,0 +1,91 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/util/af-types.h>
+#include <app/util/basic-types.h>
+#include <lib/support/Span.h>
+#include <platform/CHIPDeviceConfig.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+struct PowerSourceClusterInfo
+{
+    PowerSourceClusterInfo() : mClusterEndpoint(kInvalidEndpointId) {}
+    explicit PowerSourceClusterInfo(EndpointId powerClusterEndpointId) : mClusterEndpoint(powerClusterEndpointId) {}
+    CHIP_ERROR SetEndpointList(Span<EndpointId> endpointList)
+    {
+        if (endpointList.size() == 0)
+        {
+            mBuf          = nullptr;
+            mEndpointList = Span<EndpointId>();
+            return CHIP_NO_ERROR;
+        }
+
+        mBuf = std::make_unique<EndpointId[]>(endpointList.size());
+        if (mBuf == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        memcpy(mBuf.get(), endpointList.data(), endpointList.size() * sizeof(EndpointId));
+        mEndpointList = Span<EndpointId>(mBuf.get(), endpointList.size());
+        return CHIP_NO_ERROR;
+    }
+    EndpointId mClusterEndpoint        = kInvalidEndpointId;
+    std::unique_ptr<EndpointId[]> mBuf = nullptr;
+    Span<EndpointId> mEndpointList;
+};
+
+class PowerSourceServer
+{
+public:
+    static PowerSourceServer & Instance();
+
+    // Caller does not need to retain the span past the call point as these are copied into an internal storage
+    CHIP_ERROR SetEndpointList(EndpointId powerSourceClusterEndpoint, Span<EndpointId> endpointList);
+    // returns nullptr if there's not endpoint list set for this power source cluster endpoint id.
+    const Span<EndpointId> * GetEndpointList(EndpointId powerSourceClusterEndpoint) const;
+
+private:
+    // There's probably a good way to properly set this from zap or have this grow dynamically as dynamic endpoints are
+    // finalized, but for now, let's just say 10 power clusters is probably ok.
+    // TODO: is there a way to get a static count of endpoints? CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT?
```

Comment:
`EMBER_AF_POWER_SOURCE_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT` but you have to include the ember config header for that.  Might be better to put that in the .cpp not here.

Reply:
That was my concern too, but it needs to be in the header if I'm going to use it to size the array. So I can change it, but it does pull the gen_config header in.

Reply:
That's why I suggested putting the array in the .cpp as a file static instead of a class static.

Reply:
I think I misunderstood what you were saying - I thought you were trying to jam a constexpr into a cpp file.

Reply:
This now now moved into the cpp.